app/lib/service/email/backend.dart

```diff
@@ -0,0 +1,137 @@
+
+  EmailBackend(this._db);
+
+  /// Creates [OutgoingEmail] entities that can be stored alongside a transaction.
+  List<OutgoingEmail> prepareEntities(EmailMessage msg) {
+    return msg.recipients
+        .map(
+          (recipient) => OutgoingEmail.init(
+            fromEmail: msg.from.email,
+            recipientEmail: recipient.email,
+            subject: msg.subject,
+            bodyText: msg.bodyText,
+          ),
+        )
+        .toList();
+  }
+
+  /// Queries all [OutgoingEmail] objects and tries to send out the email,
+  /// deleting the entry after the email was sent successfully.
+  ///
+  /// The processing of the Datastore query will be stopped after [stopAfter]
+  /// duration has elapsed. This allows the periodic task to complete within
+  /// the planned time window.
+  ///
+  /// When [skipRecentEntries] is true, the newly created entries are
+  /// not sent immediately, to prevent race conditions and double-sending
+  /// of emails.
+  ///
+  /// Returns the number of successfully sent emails.
+  Future<int> trySendAllOutgoingEmails({
+    Duration? stopAfter,
+    bool skipRecentEntries = false,
+  }) async {
+    final sw = Stopwatch()..start();
+    final query = _db.query<OutgoingEmail>()..order('-created');
+    var successful = 0;
+    await for (final m in query.run()) {
+      if (stopAfter != null && sw.elapsed > stopAfter) break;
+      if (skipRecentEntries && m.age.inMinutes == 0) continue;
+      if (m.isDead) continue;
+      if (!m.mayAttemptNow) continue;
+      if (await trySendOutgoingEmail(m.uuid)) {
+        successful++;
+      }
+    }
+    return successful;
+  }
+
+  /// Tries to send emails with the given [ids]. The
+  /// [OutgoingEmail] entry will be deleted after the
+  /// email was sent successfully.
+  Future<void> trySendOutgoingEmails(Iterable<String> ids) async {
+    for (final id in ids) {
+      await trySendOutgoingEmail(id);
+    }
+  }
+
+  /// Tries to send email with the given [id]. The
+  /// [OutgoingEmail] entry will be deleted after the
+  /// email was sent successfully.
+  ///
+  /// Returns true if the email was sent successfully.
+  Future<bool> trySendOutgoingEmail(String id) async {
+    return await withRetryTransaction(_db, (tx) async {
+      final o = await tx.lookupOrNull<OutgoingEmail>(
+          _db.emptyKey.append(OutgoingEmail, id: id));
+      if (o == null || o.isDead || !o.mayAttemptNow) {
+        return false;
+      }
+      try {
+        await emailSender.sendMessage(EmailMessage(
+          EmailAddress(o.fromEmail!),
+          [EmailAddress(o.recipientEmail!)],
+          o.subject!,
+          o.bodyText!,
+        ));
+        // successful send, deleting entry
+        tx.delete(o.key);
+        return true;
+      } catch (e, st) {
+        _logger.warning('Email sending failed.', e, st);
+
+        // failed to send, retry later
+        o.attempts++;
+        o.lastAttempted = clock.now().toUtc();
+        o.lastError = e.toString();
```

Comment:
here is a problem. You don't know that this means we failed to send the email.

If we get a TCP connection failure after the thing has been sent, we risk going here.. while the email is sent.